File: DFS/UniquePathsII.js
/*
Follow up for "Unique Paths":

Now consider if some obstacles are added to the grids. How many unique paths would there be?

An obstacle and empty space is marked as 1 and 0 respectively in the grid.

For example,
There is one obstacle in the middle of a 3x3 grid as illustrated below.

[
  [0,0,0],
  [0,1,0],
  [0,0,0]
]
The total number of unique paths is 2.
*/

/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
    var dp = [];
    var n = obstacleGrid.length;
    var m = obstacleGrid[0].length;
    var row = false;
    var col = false;
    if(n <= 0 || m <= 0) return 0;
    for(var i = 0; i < n; i++){
        dp.push([]);
        if(obstacleGrid[i][0] === 0 && !row) dp[i][0] = 1;
        else {
            row = true;
            dp[i][0] = 0;
        }
    }
    for(var j = 0; j < m; j++){
        if(obstacleGrid[0][j] === 0 && !col) dp[0][j] = 1;
        else {
            col = true;
            dp[0][j] = 0;
        }
    }
    for(var l = 1; l < n; l++){
        for(var k = 1; k < m; k++){
            if(obstacleGrid[l][k] == 1) dp[l][k] = 0;
            else dp[l][k] = dp[l - 1][k] + dp[l][k - 1];
        }
    }
    return dp[n - 1][m - 1];
};

File: DFS/UniquePathsII.java
/*
Follow up for "Unique Paths":

Now consider if some obstacles are added to the grids. How many unique paths would there be?

An obstacle and empty space is marked as 1 and 0 respectively in the grid.

For example,
There is one obstacle in the middle of a 3x3 grid as illustrated below.

[
  [0,0,0],
  [0,1,0],
  [0,0,0]
]
The total number of unique paths is 2.
*/

public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int n = obstacleGrid.length;
        int m = obstacleGrid[0].length;
        if(m <= 0 || n <= 0) return 0;
        int[][] dp = new int[n][m];
        boolean col = false;
        boolean row = false;
        for(int i = 0; i < n; i++) {
            if(!col && obstacleGrid[i][0] == 0) dp[i][0] = 1;
            else {
                dp[i][0] = 0;
                col = true;
            }
        }
        for(int i = 0; i < m; i++){
            if(!row && obstacleGrid[0][i] == 0) dp[0][i] = 1;
            else {
                dp[0][i] = 0;
                row = true;
            }
        }
        for(int i = 1; i < n; i++){
            for(int j = 1; j < m; j++){ 
                if(obstacleGrid[i][j] == 1) dp[i][j] = 0;
                else dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[n - 1][m - 1];
    }
}